### PR TITLE
Add command dev:asset:clear, tasks #141

### PR DESCRIPTION
**Pull request by jthln**
Add command dev:asset:clear, tasks #141
Remove duplicate cache:list entry in readme.rst

Magerun pull-request check-list:

- [x] Pull request against develop branch (if not, just close and create a new one against it)
- [x] README.md reflects changes (if any)

Subject: Add command dev:asset:clear, tasks #141

### CODE REVIEW

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ClearCommand extends AbstractMagentoCommand
+{
+    /**
+     * @var Filesystem
+     */
+    protected $filesystem;
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+    /**
+     * @var OutputInterface
+     */
+    protected $output;
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+    /**
+     * @var string[]
+     */
+    protected $messages = [];
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    public function inject(
+        Filesystem $filesystem
+    ) {
+        $this->filesystem = $filesystem;
+    }
+
+    /**
+     * @return Filesystem
+     */
+    public function getFilesystem()
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        return $this->filesystem;
+    }
+
+    /**
+     * @return OutputInterface
+     */
+    public function getOutput()
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    {
+        return $this->output;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getMessages()
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            )
+            ->setDescription('Clear static assets')
+            ->setHelp($help);
+    }
+
+    /**
+     * @param string $code
+     * @return WriteInterface|null
+     */
+    public function getDirectoryWrite($code)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+        return $dir;
+    }
+
+    /**
+     * @param string $theme
+     * @param string $code
+     * @return string[]
+     */
+    public function findThemePaths($theme, $code)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+        return $paths;
+    }
+
+    /**
+     * @param WriteInterface $dir
+     * @param string $path
+     * @return void
+     */
+    public function deleteDirectory($dir, $path)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+                $this->messages[] = '<debug>' . $e->getTraceAsString() . '</debug>';
+            }
+        }
+    }
+
+    /**
+     * @param string $code
+     * @return void
+     */
+    public function emptyDirectory($code)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        }
+    }
+
+    /**
+     * @param string $theme
+     * @param string $code
+     * @return void
+     */
+    public function deleteThemeDirectories($theme, $code)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+                $this->deleteDirectory($dir, $path);
+            }
+        }
+    }
+
+    /**
+     * @param array $themes
+     * @return void
+     */
+    public function clearThemes($themes)
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        foreach ($themes as $theme) {
+            $this->deleteThemeDirectories($theme, DirectoryList::STATIC_VIEW);
+            $this->deleteThemeDirectories($theme, DirectoryList::TMP_MATERIALIZATION_DIR);
+        }
+    }
+
+    /**
+     * @return void
+     */
+    public function clearAllThemes()
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: Visibility: Please prefer private.

**Comment by tkn98** [comment]
We need to fix the travis build and/or need to align the PhpUnit version. Due to Magento2 dev requirements (which we hopefully don't need but just install) there is a conflict with autoloading in Magento 2 and Magerun 2 for Phpunit.

* https://travis-ci.org/netz98/n98-magerun2/jobs/202162135
* Let's just cross fingers for a momemt that issues like https://github.com/magento/magento2/issues/1690 are fixed.

**Comment by ktomk** [comment]
@jthln: Fixed #272, just FYI.

**Comment by jthln** [comment]
The PHP 5.5 with Magento 2.1.5 combo issue should be gone with your latest fixes.

Regarding the other issue:

```
1) N98\Magento\Command\Developer\Asset\ClearCommandTest::testExecuteClearTheme
UnexpectedValueException: Unable to retrieve deployment version of static files from the file system.
```

It seems, that this is caused by the preceding test `testExecuteClearAll`, which deletes `pub/static/deployed_version.txt`, but that file needs to exist for `testExecuteClearTheme`.

I've not tested it yet, but probably changing the order of the two tests could work as temporary workaround. I'd prefer a real fix, though.

I'm wondering why this happens in Magento 2.0.11, 2.0.12, 2.1.3, 2.1.4 and 2.1.5, but does not happen in Magento 2.1.0, 2.1.1 and 2.1.2, or in 2.0.10 and below.

Probably some mechanisms have changed multiple times, back and forth.

I'll take a closer look at it, but this will take some time.

**Comment by ktomk** [comment]
I'd say the error message actually makes sense. As it's in the second part that runs the test-suite against production mode, the configured theme is gone after delete (at least the static assets of it).

And that's what the command is for, right? It should be used in `developer` or `default` mode, in production mode it will actually destroy things, correct?

So a check of the mode could be suitable. I have it on the level of a test, let me fetch it for you: https://github.com/netz98/n98-magerun2/commit/f68aa3856cb3d689a3ce7b181a142fba82ca018e#diff-c7965d15834f6e47064abc71a187365bR25 - just as an example, there *might* be something to get the mode in the Magerun Application or the base command class.

The reasoning why only in these versions, I don't know but would assume Magento is just more picky nowadays as otherwise there are more weird error messages with file-fallbacks which I would guess can be the case with assets.

To cut this, if you say this is a developer command (as it looks like to me), we should forbid to run it in production mode unless this is really requested from some users. Don't know what a valid use-case would be. Perhaps explicit re-generation but after setup an application in production mode it should just run and there should be no need for "hotpatching" assets (kill and re-generate).

Your decision. The test will fail in it's current form in production mode and this is what happens. I would skip it in production mode so that it's documented that using it will break things.

- [About Magento modes (Magento2 Devdocs)](http://devdocs.magento.com/guides/v2.0/config-guide/bootstrap/magento-modes.html)

File: src/N98/Magento/Command/Developer/Asset/ClearCommand.php [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+     * @param OutputInterface $output
+     * @return void
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->detectMagento($output, true);
+        if (!$this->initMagento()) {
+            return;
+        }
+        $this->output = $output;
```
Comment (ktomk) [inline src/N98/Magento/Command/Developer/Asset/ClearCommand.php]: This needs a check that prevents running this command in production mode. This is for the users safety if they don't have production on a read-only file-system for the assets which can not be assumed. Then deleting the assets in error would require to re-generate them which can take quite some time.

**Comment by cmuench** [comment]
Thanks Jürgen.

**Comment by niclashoyer** [comment]
@cmuench @ktomk 
we really love the `dev:asset:clear` command! However we sometimes need this command on a production system to fix assets. Would it be possible to add a `-f` (force) flag as with `setup:static-content:deploy` to allow asset removing in `production` mode?

**Comment by cmuench** [comment]
@niclashoyer Sounds good. Can you create a new issue for that? So you get informations about updates on the topic.

**Comment by niclashoyer** [comment]
Sure: #337

**Pull request merged**